lib/Parser/Parse.cpp

```diff
@@ -3055,6 +3056,10 @@ ParseNodePtr Parser::ParseTerm(BOOL fAllowCall,
         {
             m_deferEllipsisError = false;
         }
+
+        // Restore the deferred ellipsis error state now that an error has been ruled out.
+        m_deferEllipsisError = deferEllipsisErrorSave;
```

Comment:
Do we also need to save/restore the RestorePoint (m_EllipsisErrorLoc) if there is one? (And only if there is one?)

Reply:
That depends how much we care about emitting the first error we find. I'm thinking maybe just the paren depth save/restore should be enough.

Reply:
Well...given that it's unusual to have one of these, it's probably a negligible cost to copy the restore point if it exists. I can imagine a compliance test failing if the wrong error is reported.

Reply:
I mean more in the case of something like (...a, (...b, (...c etc. Presumably you'd want to emit the first one and not the innermost one, which is why we currently keep the first restore point even if we find subsequent errors. The only exception is if we find another error before we get back up to the spread/rest error, in which case that will be emitted first.

Reply:
So does that mean that if ...b turned out to be an error, we would report the error at the location of ...a?

Reply:
Hmm.. We might already have that problem if we have multiple paren depth >1 potential errors. 

The easiest option seems to be reporting the innermost confirmed error, because we don't really know what we're going to get as we finish parsing those nested paren exprs. We might get to the first one we captured, but we could have found a completely different syntax error in the mean time.

We could be smarter about using more context info to error out early, but again this is a rare situation. I might see what other engines do with some weird cases.

Reply:
It's definitely more important that we report the correct location of the error we detected than it is to report the one that is lexically first, if that makes sense.